File: market-data/crypto-news/examples/javascript/streaming.js
/**
 * Real-time Streaming Examples - JavaScript/Node.js
 * Free Crypto News API - https://github.com/nirholas/free-crypto-news
 * 
 * Examples for WebSocket and Server-Sent Events (SSE) streaming.
 */

const WebSocket = require('ws');
const EventSource = require('eventsource');

const BASE_URL = 'https://cryptocurrency.cv';
const WS_URL = 'wss://cryptocurrency.cv';

// =============================================================================
// WebSocket - Real-time News Stream
// =============================================================================

/**
 * Connect to real-time news WebSocket stream.
 * @param {Object} options - Connection options
 * @param {Function} options.onMessage - Message handler
 * @param {Function} [options.onError] - Error handler
 * @param {Function} [options.onClose] - Close handler
 * @param {string[]} [options.categories] - Filter by categories
 * @returns {WebSocket} WebSocket connection
 */
function connectNewsStream({ onMessage, onError, onClose, categories } = {}) {
  const ws = new WebSocket(`${WS_URL}/api/stream/news`);
  
  ws.on('open', () => {
    console.log('📡 Connected to news stream');
    
    // Subscribe to specific categories if provided
    if (categories && categories.length > 0) {
      ws.send(JSON.stringify({
        type: 'subscribe',
        categories: categories
      }));
    }
  });
  
  ws.on('message', (data) => {
    try {
      const message = JSON.parse(data);
      if (onMessage) onMessage(message);
    } catch (e) {
      console.error('Parse error:', e);
    }
  });
  
  ws.on('error', (error) => {
    console.error('WebSocket error:', error);
    if (onError) onError(error);
  });
  
  ws.on('close', () => {
    console.log('🔌 Disconnected from news stream');
    if (onClose) onClose();
  });
  
  return ws;
}

// =============================================================================
// WebSocket - Market Data Stream
// =============================================================================

/**
 * Connect to real-time market data stream.
 * @param {Object} options - Connection options
 * @param {string[]} options.symbols - Symbols to subscribe to
 * @param {Function} options.onTick - Tick handler
 * @returns {WebSocket} WebSocket connection
 */
function connectMarketStream({ symbols, onTick } = {}) {
  const ws = new WebSocket(`${WS_URL}/api/stream/market`);
  
  ws.on('open', () => {
    console.log('📈 Connected to market stream');
    
    // Subscribe to symbols
    if (symbols && symbols.length > 0) {
      ws.send(JSON.stringify({
        type: 'subscribe',
        symbols: symbols
      }));
    }
  });
  
  ws.on('message', (data) => {
    try {
      const tick = JSON.parse(data);
      if (onTick) onTick(tick);
    } catch (e) {
      console.error('Parse error:', e);
    }
  });
  
  ws.on('error', console.error);
  
  return ws;
}

// =============================================================================
// WebSocket - Whale Alerts Stream
// =============================================================================

/**
 * Connect to real-time whale alerts stream.
 * @param {Object} options - Connection options
 * @param {number} [options.minValue=1000000] - Minimum transaction value
 * @param {Function} options.onAlert - Alert handler
 * @returns {WebSocket} WebSocket connection
 */
function connectWhaleStream({ minValue = 1000000, onAlert } = {}) {
  const ws = new WebSocket(`${WS_URL}/api/stream/whales`);
  
  ws.on('open', () => {
    console.log('🐋 Connected to whale alerts stream');
    
    ws.send(JSON.stringify({
      type: 'config',
      minValue: minValue
    }));
  });
  
  ws.on('message', (data) => {
    try {
      const alert = JSON.parse(data);
      if (onAlert) onAlert(alert);
    } catch (e) {
      console.error('Parse error:', e);
    }
  });
  
  return ws;
}

// =============================================================================
// WebSocket - Sentiment Stream
// =============================================================================

/**
 * Connect to real-time sentiment updates stream.
 * @param {Object} options - Connection options
 * @param {string[]} [options.assets] - Assets to track
 * @param {Function} options.onUpdate - Update handler
 * @returns {WebSocket} WebSocket connection
 */
function connectSentimentStream({ assets, onUpdate } = {}) {
  const ws = new WebSocket(`${WS_URL}/api/stream/sentiment`);
  
  ws.on('open', () => {
    console.log('😊 Connected to sentiment stream');
    
    if (assets && assets.length > 0) {
      ws.send(JSON.stringify({
        type: 'subscribe',
        assets: assets
      }));
    }
  });
  
  ws.on('message', (data) => {
    try {
      const update = JSON.parse(data);
      if (onUpdate) onUpdate(update);
    } catch (e) {
      console.error('Parse error:', e);
    }
  });
  
  return ws;
}

// =============================================================================
// SSE - Server-Sent Events for News
// =============================================================================

/**
 * Connect to SSE news stream.
 * @param {Object} options - Connection options
 * @param {string} [options.category] - Filter by category
 * @param {Function} options.onNews - News handler
 * @returns {EventSource} EventSource connection
 */
function connectNewsSSE({ category, onNews } = {}) {
  const params = category ? `?category=${category}` : '';
  const es = new EventSource(`${BASE_URL}/api/sse/news${params}`);
  
  es.onopen = () => {
    console.log('📡 SSE Connected to news stream');
  };
  
  es.onmessage = (event) => {
    try {
      const news = JSON.parse(event.data);
      if (onNews) onNews(news);
    } catch (e) {
      console.error('Parse error:', e);
    }
  };
  
  es.onerror = (error) => {
    console.error('SSE Error:', error);
  };
  
  return es;
}

// =============================================================================
// SSE - Order Book Stream
// =============================================================================

/**
 * Connect to SSE order book stream.
 * @param {string} symbol - Trading pair
 * @param {Function} onUpdate - Update handler
 * @returns {EventSource} EventSource connection
 */
function connectOrderbookSSE(symbol = 'BTCUSDT', onUpdate) {
  const es = new EventSource(`${BASE_URL}/api/orderbook/stream?symbol=${symbol}`);
  
  es.onmessage = (event) => {
    try {
      const update = JSON.parse(event.data);
      if (onUpdate) onUpdate(update);
    } catch (e) {
      console.error('Parse error:', e);
    }
  };
  
  return es;
}

// =============================================================================
// SSE - Breaking News Stream
// =============================================================================

/**
 * Connect to SSE breaking news stream.
 * @param {Function} onBreaking - Breaking news handler
 * @returns {EventSource} EventSource connection
 */
function connectBreakingSSE(onBreaking) {
  const es = new EventSource(`${BASE_URL}/api/sse/breaking`);
  
  es.onmessage = (event) => {
    try {
      const breaking = JSON.parse(event.data);
      if (onBreaking) onBreaking(breaking);
    } catch (e) {
      console.error('Parse error:', e);
    }
  };
  
  return es;
}

// =============================================================================
// EXPORTS
// =============================================================================

module.exports = {
  // WebSocket streams
  connectNewsStream,
  connectMarketStream,
  connectWhaleStream,
  connectSentimentStream,
  
  // SSE streams
  connectNewsSSE,
  connectOrderbookSSE,
  connectBreakingSSE,
  
  BASE_URL,
  WS_URL
};

// =============================================================================
// EXAMPLE USAGE
// =============================================================================

async function runExamples() {
  console.log('\n' + '='.repeat(60));
  console.log('FREE CRYPTO NEWS API - STREAMING EXAMPLES');
  console.log('='.repeat(60));
  
  // Example 1: News Stream
  console.log('\n📰 1. Real-time News Stream (5 seconds)');
  const newsWs = connectNewsStream({
    categories: ['bitcoin', 'ethereum'],
    onMessage: (news) => {
      console.log(`   New article: ${news.title?.substring(0, 50)}...`);
    }
  });
  
  // Example 2: Market Stream
  console.log('\n📈 2. Market Data Stream (5 seconds)');
  const marketWs = connectMarketStream({
    symbols: ['BTC', 'ETH', 'SOL'],
    onTick: (tick) => {
      console.log(`   ${tick.symbol}: $${tick.price}`);
    }
  });
  
  // Example 3: Whale Alerts
  console.log('\n🐋 3. Whale Alerts Stream (5 seconds)');
  const whaleWs = connectWhaleStream({
    minValue: 5000000,
    onAlert: (alert) => {
      console.log(`   Whale: ${alert.amount} ${alert.asset} ($${alert.value})`);
    }
  });
  
  // Close after 5 seconds
  setTimeout(() => {
    console.log('\n🔌 Closing connections...');
    newsWs.close();
    marketWs.close();
    whaleWs.close();
    
    console.log('\n' + '='.repeat(60));
    console.log('Streaming examples completed!');
    console.log('='.repeat(60));
    process.exit(0);
  }, 5000);
}

if (require.main === module) {
  runExamples().catch(console.error);
}

// =============================================================================
// ADVANCED EXAMPLES
// =============================================================================

/**
 * Example: Trading Bot with Real-time Data
 */
class TradingBot {
  constructor(config) {
    this.config = config;
    this.streams = {};
  }
  
  start() {
    // Connect to market data
    this.streams.market = connectMarketStream({
      symbols: this.config.symbols,
      onTick: this.handleTick.bind(this)
    });
    
    // Connect to news for sentiment
    this.streams.news = connectNewsStream({
      categories: this.config.categories,
      onMessage: this.handleNews.bind(this)
    });
    
    // Connect to whale alerts
    this.streams.whales = connectWhaleStream({
      minValue: this.config.minWhaleValue,
      onAlert: this.handleWhaleAlert.bind(this)
    });
    
    console.log('🤖 Trading bot started');
  }
  
  handleTick(tick) {
    // Implement your trading logic
    console.log(`Tick: ${tick.symbol} = $${tick.price}`);
  }
  
  handleNews(news) {
    // React to news
    console.log(`News: ${news.title?.substring(0, 50)}`);
  }
  
  handleWhaleAlert(alert) {
    // React to whale movements
    console.log(`🐋 Whale Alert: ${alert.amount} ${alert.asset}`);
  }
  
  stop() {
    Object.values(this.streams).forEach(stream => stream.close());
    console.log('🤖 Trading bot stopped');
  }
}

/**
 * Example: News Aggregator with SSE
 */
function createNewsAggregator(categories, onArticle) {
  const streams = {};
  
  categories.forEach(category => {
    streams[category] = connectNewsSSE({
      category,
      onNews: (article) => {
        onArticle({ category, article });
      }
    });
  });
  
  return {
    streams,
    close: () => {
      Object.values(streams).forEach(stream => stream.close());
    }
  };
}

module.exports.TradingBot = TradingBot;
module.exports.createNewsAggregator = createNewsAggregator;

File: market-data/crypto-news/examples/rust/streaming.rs
//! Real-Time Streaming Example
//!
//! Demonstrates WebSocket streaming for live news and price updates.
//!
//! Run: cargo run --example streaming --features websocket

use fcn_sdk::{Client, Result, StreamEvent};
use futures_util::StreamExt;
use std::time::Duration;
use tokio::time::timeout;

#[tokio::main]
async fn main() -> Result<()> {
    let client = Client::new(None)?;

    println!("═══════════════════════════════════════════════════════════");
    println!("          Real-Time Crypto Streaming - Rust");
    println!("═══════════════════════════════════════════════════════════\n");

    // 1. Stream Live News
    println!("📰 Connecting to live news stream...");
    println!("───────────────────────────────────────────────────────────");
    
    let mut news_stream = client.stream_news().await?;
    println!("✅ Connected! Waiting for news (10 second timeout)...\n");

    let news_result = timeout(Duration::from_secs(10), async {
        let mut count = 0;
        while let Some(result) = news_stream.next().await {
            match result {
                Ok(article) => {
                    count += 1;
                    println!("📰 [{}] {}", article.source, article.title);
                    if let Some(tickers) = &article.tickers {
                        if !tickers.is_empty() {
                            println!("   💹 {}", tickers.join(", "));
                        }
                    }
                    println!();
                    
                    if count >= 3 {
                        break;
                    }
                }
                Err(e) => {
                    eprintln!("   ❌ Error: {}", e);
                }
            }
        }
        count
    })
    .await;

    match news_result {
        Ok(count) => println!("   Received {} articles\n", count),
        Err(_) => println!("   ⏱️  Timeout - no news in last 10 seconds\n"),
    }

    // 2. Stream Live Prices
    println!("💰 Connecting to price stream...");
    println!("───────────────────────────────────────────────────────────");

    let symbols = vec!["BTC", "ETH", "SOL", "XRP", "DOGE"];
    let mut price_stream = client.stream_prices(&symbols).await?;
    println!("✅ Streaming: {}\n", symbols.join(", "));

    let price_result = timeout(Duration::from_secs(10), async {
        let mut updates = 0;
        while let Some(result) = price_stream.next().await {
            match result {
                Ok(price) => {
                    updates += 1;
                    let change_emoji = if price.change_24h >= 0.0 { "📈" } else { "📉" };
                    println!(
                        "   {} {} ${:.2} ({:+.2}%)",
                        change_emoji,
                        price.symbol,
                        price.price,
                        price.change_24h
                    );
                    
                    if updates >= 10 {
                        break;
                    }
                }
                Err(e) => {
                    eprintln!("   ❌ Error: {}", e);
                }
            }
        }
        updates
    })
    .await;

    match price_result {
        Ok(count) => println!("\n   Received {} price updates\n", count),
        Err(_) => println!("\n   ⏱️  Timeout\n"),
    }

    // 3. Combined Event Stream
    println!("🔄 Combined event stream (news + prices + alerts):");
    println!("───────────────────────────────────────────────────────────");

    let mut event_stream = client.stream_all().await?;
    println!("✅ Listening for all events...\n");

    let event_result = timeout(Duration::from_secs(15), async {
        let mut events = 0;
        while let Some(result) = event_stream.next().await {
            match result {
                Ok(event) => {
                    events += 1;
                    match event {
                        StreamEvent::News(article) => {
                            println!("   📰 NEWS: {}", article.title);
                        }
                        StreamEvent::Price(price) => {
                            println!("   💰 PRICE: {} ${:.2}", price.symbol, price.price);
                        }
                        StreamEvent::WhaleAlert(alert) => {
                            println!(
                                "   🐋 WHALE: {} {} ${:.0}",
                                alert.symbol, alert.amount, alert.usd_value
                            );
                        }
                        StreamEvent::Liquidation(liq) => {
                            println!(
                                "   💥 LIQ: {} {} ${:.0}",
                                liq.side.to_uppercase(),
                                liq.symbol,
                                liq.quantity
                            );
                        }
                        StreamEvent::Signal(signal) => {
                            println!(
                                "   📊 SIGNAL: {} {} @ ${:.2}",
                                signal.action.to_uppercase(),
                                signal.symbol,
                                signal.price
                            );
                        }
                    }
                    
                    if events >= 10 {
                        break;
                    }
                }
                Err(e) => {
                    eprintln!("   ❌ Error: {}", e);
                }
            }
        }
        events
    })
    .await;

    match event_result {
        Ok(count) => println!("\n   Received {} events\n", count),
        Err(_) => println!("\n   ⏱️  Timeout\n"),
    }

    // 4. Ticker-Specific Stream
    println!("🎯 Streaming Bitcoin-specific events:");
    println!("───────────────────────────────────────────────────────────");

    let mut btc_stream = client.stream_ticker("BTC").await?;
    println!("✅ Listening for BTC events...\n");

    let btc_result = timeout(Duration::from_secs(10), async {
        let mut events = 0;
        while let Some(result) = btc_stream.next().await {
            if let Ok(event) = result {
                events += 1;
                println!("   🔶 BTC Event: {:?}", event);
                if events >= 5 {
                    break;
                }
            }
        }
        events
    })
    .await;

    match btc_result {
        Ok(count) => println!("\n   Received {} BTC events\n", count),
        Err(_) => println!("\n   ⏱️  Timeout\n"),
    }

    println!("═══════════════════════════════════════════════════════════");
    println!("              Streaming Demo Complete! 🔴");
    println!("═══════════════════════════════════════════════════════════");

    Ok(())
}
